include/ideep/operators/conv.hpp

```diff
@@ -1838,20 +1838,25 @@ struct convolution_forward
     if (reorder_src) {
       auto expected_dst_desc = param.pd.dst_desc();
       tensor expected_dst;
-      // dst not init in FW or has same desc with expected desc.
-      if (dst.is_empty() || dst.get_desc() == expected_dst_desc) {
-        dst.reinit_if_possible(expected_dst_desc);
-        expected_dst = dst;
-      } else {
+      if (dst.is_empty() || dst.get_desc() != expected_dst_desc){
+        // If dst buffer are not given by user or user given dst buffer are not under expected format
+        // We need init a new one
         expected_dst.init(expected_dst_desc);
-        if (param.op_attr.has_op_kind(kind::sum)) {
+        if (!dst.is_empty() && param.op_attr.has_op_kind(kind::sum)) {
+          // We need copy the content of given buffer if matmul is fused with sum
           expected_dst.feed_from(dst);
         }
+      } else {
+        // The format of given dst buffer is expected
+        expected_dst = dst;
       }
       args.insert({DNNL_ARG_DST, expected_dst});
       primitive.execute(stream::default_stream(), args);
-      // dst has been init in FW side, but has diff desc with expected_dst.
-      if (dst.get_desc() != expected_dst.get_desc()) {
+      // reorder back to dst's buffer if needed
+      if (dst.is_empty() || dst.get_desc() == expected_dst.get_desc() ||
+          !dst.get_desc().has_same_shape_as(expected_dst.get_desc())) {
+        dst = expected_dst;
+      } else {
```

Comment:
As long as dst is not empty and dst does not share the buffer with expected_dst, we do feed_from call. Is this correct and simpler?

Reply:
Hi @XiaobingSuper, could you comment? Can the conditions cover all cases?

Reply:
> As long as dst is not empty and dst does not share the buffer with expected_dst, we do feed_from call. Is this correct and simpler?

No. I think we also need to consider their shapes.
If dst is not empty and it does not share the buffer with expected_dst and it does not have the same shape as expected, we cannot do `feed_from`.

Reply:
How do we handle the case where user provides the dst buffer but the shape is different from expected_dst? In that case, expected_dst would have different buffer from dst? Should we copy the content to the user provided buffer in dst?

Reply:
I can think of two ways:
- Reinit dst buffer so that it has the expected shape. Then feed from expected_dst.
- Dst = expected_dst

Here we do it in the second way.

Reply:
> Here we do it in the second way.

Is this correct that you don't write the content into user-provided buffer?

Reply:
I think the problem is the size of dst buffer is not correct. We cannot write it directly. In both ways we create a new buffer for it.
Hi @XiaobingSuper, do you have more comments?

Reply:
If the user provides a buffer with the correct shape: There is no doubt we should write the content.
If the user provides a buffer with the wrong shape: We cannot write into the buffer and need re-init. But this "re-init" is taken place in "ideep", can you confirm whether this re-inited buffer will be freed when it is returned to aten/caffe2.

Reply:
Maybe caffe2 has some ways to keep the buffer allocated in ideep from being freed. But for aten, I guess we should write to the buffer given by aten user ( And aten user should not give a buffer with wrong shape),

Reply:
> Thanks @zhuhaozhe I saw that `ideep::tensor::buffer` is a shared_ptr, and `reinit` will create a new buffer and `operator=` will assign the shared_ptr. So, I think the buffer remains after the call. Anyway, in previous implementation, dst is always `reinit`'ed if dst's shape is not expected. So, it will not cause regression from this point of view. Hi @jgong5 Do you think it makes sense?

The new buffer is created by ideep scope. How do you convert ideep tensor to aten tensor? Do they share buffer? If they share buffer (which means there is no copy taken place while converting ideep tensor to aten tensor), while ideep tensor is dead, the buffer will be freed. (The aten scope do not aware ideep buffer and cannot hold it).

Reply:
> > Thanks @zhuhaozhe I saw that `ideep::tensor::buffer` is a shared_ptr, and `reinit` will create a new buffer and `operator=` will assign the shared_ptr. So, I think the buffer remains after the call. Anyway, in previous implementation, dst is always `reinit`'ed if dst's shape is not expected. So, it will not cause regression from this point of view. Hi @jgong5 Do you think it makes sense?
> 
> The new buffer is created by ideep scope. How do you convert ideep tensor to aten tensor? Do they share buffer? If they share buffer (which means there is no copy taken place while converting ideep tensor to aten tensor), while ideep tensor is dead, the buffer will be freed. (The aten scope do not aware ideep buffer and cannot hold it).

You mean it's a problem, right? If so, what was in your mind when you added these codes then? https://github.com/intel/ideep/blob/77d662b313a762e82b389d3fd965e0098f12cd99/include/ideep/operators/matmul.hpp#L681

Anyway, I saw Caffe2 is using ideep's tensor in the caller here: https://github.com/pytorch/pytorch/blob/7a9ab5c232f54430704456d18a22f99838489817/caffe2/ideep/operators/quantization/int8_conv_op.cc#L37
I am not sure if there are conversions between ATen tensor and ideep tensor elsewhere.
And as I have said, in previous implementations, dst is always `reinit`'ed if the shape is not correct, which means a new buffer is always created in ideep in such cases. If that is correct, the new implementation should not be a problem.

Reply:
@Xia-Weiwen may I know in what case the given dst is not empty and has the different shape as the expected_dst?

Reply:
> > > Thanks @zhuhaozhe I saw that `ideep::tensor::buffer` is a shared_ptr, and `reinit` will create a new buffer and `operator=` will assign the shared_ptr. So, I think the buffer remains after the call. Anyway, in previous implementation, dst is always `reinit`'ed if dst's shape is not expected. So, it will not cause regression from this point of view. Hi @jgong5 Do you think it makes sense?
> > 
> > 
> > The new buffer is created by ideep scope. How do you convert ideep tensor to aten tensor? Do they share buffer? If they share buffer (which means there is no copy taken place while converting ideep tensor to aten tensor), while ideep tensor is dead, the buffer will be freed. (The aten scope do not aware ideep buffer and cannot hold it).
> 
> You mean it's a problem, right? If so, what was in your mind when you added these codes then?
> 
> https://github.com/intel/ideep/blob/77d662b313a762e82b389d3fd965e0098f12cd99/include/ideep/operators/matmul.hpp#L681
> 
> Anyway, I saw Caffe2 is using ideep's tensor in the caller here: https://github.com/pytorch/pytorch/blob/7a9ab5c232f54430704456d18a22f99838489817/caffe2/ideep/operators/quantization/int8_conv_op.cc#L37 I am not sure if there are conversions between ATen tensor and ideep tensor elsewhere. And as I have said, in previous implementations, dst is always `reinit`'ed if the shape is not correct, which means a new buffer is always created in ideep in such cases. If that is correct, the new implementation should not be a problem.

It will be a problem if both conditions happen:
 - (1) Ideep tenspor / Aten tensor share buffer.
 - (2) ATen pass a buffer with wrong dst buffer to ideep.

There might be no such case for matmul before, can you help to confirm will the above two cases will be satisfied at the same time for conv now?

Reply:
Have you tried stock pytorch CI with this PR? If it passed, I guess there is no such case for Conv now.

Reply:
> @Xia-Weiwen may I know in what case the given dst is not empty and has the different shape as the expected_dst?

It's found by Caffe2's UT:
https://github.com/pytorch/pytorch/blob/master/caffe2/python/ideep/convfusion_op_test.py

Reply:
Hi @jgong5:
@Guobing-Chen and I think we are probably talking about two issues here. The first is the regression (UT failure). This PR is aimed at fixing it. The second is the buffer issue. This issue is not introduced by ideep convergence. It is a long-existing issue. And fortunately, no problems found so far.
So, we may focus on the first issue first, which is blocking ideep update in stock PyTorch. For the second one, we may have discussions later. What do you think?

Reply:
OK. Let's create an issue to track this and fix in a separate PR if needed. Sounds good to you?

Reply:
> OK. Let's create an issue to track this and fix in a separate PR if needed. Sounds good to you?

OK Sounds good. Thanks.

Reply:
Hi @jgong5 I created the JIRA here: https://jira.devtools.intel.com/browse/IPB-1272
For this PR, do you have more comments?